lib/api/objectPutPart.js

```diff
@@ -72,95 +36,165 @@ function doStore(objectKeyContext, cipherBundle, request, log,
+                log.debug('access denied for user on bucket', { requestType });
+                return next(errors.AccessDenied);
             }
-            return services.getMPUBucket(
-                bucket, bucketName, log,
-                (err, mpuBucket) => {
-                    if (err) {
-                        return cb(err);
-                    }
-                    // BACKWARD: Remove to remove the old splitter
-                    if (mpuBucket.getMdBucketModelVersion() < 2) {
-                        metadataValMPUparams.splitter = constants.oldSplitter;
-                    }
-                    // We pad the partNumbers so that the parts will be sorted
-                    // in numerical order
-                    const paddedPartNumber = `000000${partNumber}`.substr(-5);
-                    const serverSideEncryption =
-                              bucket.getServerSideEncryption();
-                    if (serverSideEncryption) {
-                        return kms.createCipherBundle(
-                            serverSideEncryption, log, (err, cipherBundle) => {
-                                services.metadataValidateMultipart(
-                                    metadataValMPUparams,
-                                    doStore(objectKeyContext, cipherBundle,
-                                            request, log, paddedPartNumber,
-                                            size, uploadId,
-                                            streamingV4Params, cb));
-                            });
-                    }
-                    return services.metadataValidateMultipart(
-                        metadataValMPUparams,
-                        doStore(objectKeyContext, null, request, log,
-                                paddedPartNumber, size, uploadId,
-                                streamingV4Params, cb));
+
+            return next(null, bucket);
+        },
+        // Get bucket server-side encryption, if it exists.
+        (bucket, next) => {
+            const encryption = bucket.getServerSideEncryption();
+
+            if (encryption) {
+                return kms.createCipherBundle(encryption, log, (err, res) => {
+                    cipherBundle = res;
+                    return next(err);
                 });
-        });
+            }
+
+            return next();
+        },
+        // Get the MPU shadow bucket.
+        next => metadata.getBucket(mpuBucketName, log, (err, mpuBucket) => {
+            if (err) {
+                return next(errors.NoSuchUpload);
+            }
+            return next(null, mpuBucket);
+        }),
+        // Check authorization of the MPU shadow bucket.
+        (mpuBucket, next) => {
+            // BACKWARD: Remove to remove the old splitter
+            if (mpuBucket.getMdBucketModelVersion() < 2) {
+                splitter = constants.oldSplitter;
+            }
+
+            const args = {
```

Comment:
Instead of doing a listing, I think we can do a straight get object from metadata, can't we?  At some point we had more info in the key so we weren't able to do that but I think it's possible now and would be more straightforward than doing a listing to get one key.